File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/Diagnostic.scala
package org.enso.compiler.core.ir

/** A representation of various kinds of diagnostic in the IR. */
trait Diagnostic extends Serializable {

  /** @param source Location of the diagnostic.
    * @return a human-readable description of this error condition.
    */
  def message(source: (IdentifiedLocation => String)): String

  /** @param source Location of the diagnostic.
    * @return a human-readable description of this error condition, formatted for immediate reporting.
    */
  def formattedMessage(source: (IdentifiedLocation => String)): String =
    message(source)

  /** The location at which the diagnostic occurs. */
  val location: Option[IdentifiedLocation]

  /** The important keys identifying identity of the diagnostic
    */
  def diagnosticKeys(): Array[Any]
}

object Diagnostic {

  /** Represents the various kinds of diagnostics in the IR. */
  sealed trait Kind

  object Kind {

    /** Diagnostics that should be reported during the static compilation
      * phase of execution.
      */
    trait Static extends Kind

    /** Diagnostics that should remain at runtime for display during
      * interactive execution.
      */
    trait Interactive extends Kind
  }
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/Diagnostic.java
package org.enso.compiler.core.ir;

import scala.Function1;
import scala.Option;

/** A representation of various kinds of diagnostic in the IR. */
public interface Diagnostic {

  /**
   * @param source Location of the diagnostic.
   * @return a human-readable description of this error condition.
   */
  String message(Function1<IdentifiedLocation, String> source);

  /**
   * @param source Location of the diagnostic.
   * @return a human-readable description of this error condition, formatted for immediate
   *     reporting.
   */
  default String formattedMessage(Function1<IdentifiedLocation, String> source) {
    return message(source);
  }

  /** The location at which the diagnostic occurs. */
  IdentifiedLocation identifiedLocation();

  /** The location at which the diagnostic occurs. */
  default Option<IdentifiedLocation> location() {
    return Option.apply(identifiedLocation());
  }

  /** The important keys identifying identity of the diagnostic */
  Object[] diagnosticKeys();

  /** Represents the various kinds of diagnostics in the IR. */
  sealed interface Kind permits Kind.Static, Kind.Interactive {

    /** Diagnostics that should be reported during the static compilation phase of execution. */
    non-sealed interface Static extends Kind {}

    /** Diagnostics that should remain at runtime for display during interactive execution. */
    non-sealed interface Interactive extends Kind {}
  }
}
